core/embed/rust/src/micropython/buffer.rs

```diff
@@ -1,151 +1,61 @@
-use core::{
-    convert::TryFrom,
-    ops::{Deref, DerefMut},
-    ptr, slice, str,
-};
+use core::{convert::TryFrom, ops::Deref, ptr, slice, str};
 
 use crate::{error::Error, micropython::obj::Obj};
 
 use super::ffi;
 
-/// Represents an immutable slice of bytes stored on the MicroPython heap and
-/// owned by values that obey the `MP_BUFFER_READ` buffer protocol, such as
-/// `bytes`, `str`, `bytearray` or `memoryview`.
+/// Represents an immutable UTF-8 string managed by MicroPython GC.
+/// This either means static data, or a valid GC object.
 ///
 /// # Safety
 ///
-/// In most cases, it is unsound to store `Buffer` values in a GC-unreachable
-/// location, such as static data. It is also unsound to let the contents be
-/// modified while a reference to them is being held.
-pub struct Buffer {
+/// We assume that MicroPython is handling strings according to Python
+/// semantics, i.e., that string data is immutable.
+/// Furthermore, we assume that string data is always either static or a GC head
+/// pointer, i.e., we can never obtain a pointer into the middle of a GC string.
+///
+/// Given the above assumptions about MicroPython strings, working with
+/// StrBuffers in Rust is safe.
+pub struct StrBuffer {
     ptr: *const u8,
     len: usize,
 }
 
-impl Buffer {
+impl StrBuffer {
     pub fn empty() -> Self {
-        Self::from(b"")
-    }
-}
-
-impl TryFrom<Obj> for Buffer {
-    type Error = Error;
-
-    fn try_from(obj: Obj) -> Result<Self, Self::Error> {
-        let bufinfo = get_buffer_info(obj, ffi::MP_BUFFER_READ)?;
-
-        Ok(Self {
-            ptr: bufinfo.buf as _,
-            len: bufinfo.len as _,
-        })
-    }
-}
-
-impl Default for Buffer {
-    fn default() -> Self {
-        Self::empty()
-    }
-}
-
-impl Deref for Buffer {
-    type Target = [u8];
-
-    fn deref(&self) -> &Self::Target {
-        self.as_ref()
-    }
-}
-
-impl AsRef<[u8]> for Buffer {
-    fn as_ref(&self) -> &[u8] {
-        buffer_as_ref(self.ptr, self.len)
+        Self::from("")
     }
-}
 
-impl From<&'static [u8]> for Buffer {
-    fn from(val: &'static [u8]) -> Self {
-        Buffer {
-            ptr: val.as_ptr(),
-            len: val.len(),
+    pub fn alloc(val: &str) -> Result<Self, Error> {
+        // SAFETY:
+        // We assume that if `gc_alloc` returns successfully, the result is a valid
+        // pointer to GC-controlled memory of at least `val.len()` bytes.
+        unsafe {
+            let raw = ffi::gc_alloc(val.len(), 0) as *mut u8;
```

Comment:
I'm not sure now, but IIRC uPy has some assumptions about strings being zero-terminated?

Reply:
True, but this string will never go back to uPy (except through `TryFrom<&str> for Obj` which makes a copy)

Reply:
that said, allocating +1 and writing a null terminator is cheap and makes it less error prone for any future modifications so :woman_shrugging:

Reply:
done in 79da7b3